From a8a4691e83bb46397c858adc62c65970b5b8c9ff Mon Sep 17 00:00:00 2001
From: Stephen Yuan Jiang <syuanjiangdev@gmail.com>
Date: Sat, 10 Dec 2016 15:05:00 -0800
Subject: [PATCH 283/362] HBASE-17023 Region left unassigned due to AM and SSH
 each thinking others would do the assignment work
 (Stephen Yuan Jiang)

(cherry picked from commit b8822633b674fccb0a60d14b2b07937fb7a358e4)

Change-Id: I1c939430d409c3206ee66247f42f8cc387ad4489
Author: Balazs Meszaros
Reason: Bug
Ref: CDH-50894
---
 .../hadoop/hbase/master/AssignmentManager.java     |    3 +--
 .../apache/hadoop/hbase/master/RegionStates.java   |   25 ++++++++++++++++++++
 .../hbase/master/handler/ClosedRegionHandler.java  |    3 +--
 3 files changed, 27 insertions(+), 4 deletions(-)

diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/AssignmentManager.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/AssignmentManager.java
index f737bbd..cee0fd2 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/AssignmentManager.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/AssignmentManager.java
@@ -837,8 +837,7 @@ public class AssignmentManager extends ZooKeeperListener {
       case RS_ZK_REGION_CLOSED:
       case RS_ZK_REGION_FAILED_OPEN:
         // Region is closed, insert into RIT and handle it
-        regionStates.setLastRegionServerOfRegion(sn, encodedName);
-        regionStates.updateRegionState(regionInfo, State.CLOSED, sn);
+        regionStates.setRegionStateTOCLOSED(regionInfo, sn);
         if (!replicasToClose.contains(regionInfo)) {
           invokeAssign(regionInfo);
         } else {
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionStates.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionStates.java
index ee2cfb6..c209d35 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionStates.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionStates.java
@@ -377,6 +377,31 @@ public class RegionStates {
   }
 
   /**
+   * Set the region state to CLOSED
+   */
+  public RegionState setRegionStateTOCLOSED(
+      final HRegionInfo regionInfo,
+      final ServerName serverName) {
+    ServerName sn = serverName;
+    if (sn == null) {
+      RegionState regionState = getRegionState(regionInfo.getEncodedName());
+      if (regionState != null) {
+        sn = regionState.getServerName();
+      }
+      // TODO: if sn is null, should we dig into
+      // lastAssignments.get(regionInfo.getEncodedName() to get the server name?
+      // For now, I just keep the same logic that works in the past
+    }
+    // We have to make sure that the last region server is set to be the same as the
+    // current RS.  If we don't do that, we could run into situation that both AM and SSH
+    // think other would do the assignment work; at the end, neither does the work and
+    // region remains RIT.
+    // See HBASE-13330 and HBASE-17023
+    setLastRegionServerOfRegion(sn, regionInfo.getEncodedName());
+    return updateRegionState(regionInfo, State.CLOSED, sn);
+  }
+
+  /**
    * Update a region state. It will be put in transition if not already there.
    */
   public RegionState updateRegionState(
diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/handler/ClosedRegionHandler.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/handler/ClosedRegionHandler.java
index e0e4ee4..74b45bc 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/handler/ClosedRegionHandler.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/handler/ClosedRegionHandler.java
@@ -99,8 +99,7 @@ public class ClosedRegionHandler extends EventHandler implements TotesHRegionInf
       return;
     }
     // ZK Node is in CLOSED state, assign it.
-    assignmentManager.getRegionStates().updateRegionState(
-      regionInfo, RegionState.State.CLOSED);
+    assignmentManager.getRegionStates().setRegionStateTOCLOSED(regionInfo, null);
     // This below has to do w/ online enable/disable of a table
     assignmentManager.removeClosedRegion(regionInfo);
     assignmentManager.assign(regionInfo, true);
-- 
1.7.9.5

